feat: IMAP Integration for 2FA [BREAKING]

Skribb11es opened the pull request:
* Added `IMAP` class to handle `IMAP` requests
* Updated `FarmThread.py` by adding stats, and new exception handling for IMAP events
* Added `valid` stat to `Stats` class to allow for a thread to be forcefully closed instead of being loop restarted if need be.
* Created new exception `InvalidIMAPCredentials` for issues with IMAP credentials
* Added `IMAP` to `Browser`, along with giving it access to `Stats` for the `updateStatus` function
* Updated `Config` to support new `IMAP` fields
* Updated `confighelper` to include the optional `IMAP` fields

(Forgot syncing forked repo's removed PRs, my b)

Skribb11es committed:
IMAP Integration

Skribb11es committed:
IMAP Integration

Skribb11es committed:
IMAP Integration

Skribb11es committed:
IMAP Integration

Skribb11es committed:
IMAP Integration

Skribb11es committed:
Minor changes to match current master branch

Skribb11es committed:
Updated to latest build

14ROVI commented on src/Browser.py:
```diff
@@ -61,15 +66,35 @@ def login(self, username: str, password: str, refreshLock) -> bool:
+                    #Handles all IMAP requests
+                    try:
+                        M = imaplib2.IMAP4_SSL(imapserver)
+                        M.login(imapusername, "imappassword")
```
i think this should not have quotes around `imappassword`

Skribb11es replied:
Yea, I fixed that in my local build but forgot to push it to GitHub, will be done ASAP.

Skribb11es committed:
Fixed quoted out variable

Skribb11es commented:
Fixed the " " out var, the rest should be all good.

LeagueOfPoro commented:
You need to run `pipenv install imaplib2` and then include the updated `Pipfile` in you PR.

LeagueOfPoro commented on src/Config.py:
```diff
@@ -1,91 +1,98 @@
+                        "password": accs[account]["password"],
+                        
+                        #IMAP data
+                        "imapusername": accs[account].get("imapusername", ""),
+                        "imappassword": accs[account].get("imappassword", ""),
+                        "imapserver": accs[account].get("imapserver", ""),
```
Use camelCase to keep the config consistent. E.g., `imapUsername`. 
Also, right now, it looks like the code would overwrite the whole class on merge. There are changes in master which are not reflected here.

LeagueOfPoro commented:
Not working for me. Can somebody else test it? I am using GMail
![image](https://user-images.githubusercontent.com/95635582/218028983-4c1095d6-860d-40a8-a3a1-112bf01595b2.png)

Skribb11es commented:
Make sure you are using an App password, and not your Google account's password, for more info check [here](https://support.google.com/mail/answer/7126229?hl=en#zippy=%2Ci-cant-sign-in-to-my-email-client), also willchange to cammelCase, I normally use cammel but it seemed that most everything else was all lowercase, and will also update to the latest `config.py` file

Skribb11es committed:
Changed case from lower to cammelCase

Skribb11es committed:
Update to accomidate for case change

Skribb11es committed:
Add imaplib2

Skribb11es commented:
Should all be up to date / fixed

LeagueOfPoro commented:
> Make sure you are using an App password, and not your Google account's password,

That will be it. Will try.

Skribb11es committed:
Resolved verion issue

Skribb11es committed:
Merge branch 'master' into master

Skribb11es committed:
Delete Logger.py

Skribb11es committed:
Resolve version issue

Skribb11es commented:
Can someone elaborate on why the [BREAKING] was added to the title? As far as I'm aware, I cant find any conflicts with this and the current `master` branch...

Skribb11es commented:
Im just going to "rewrite" this tomorrow, something is seriously wrong and I have 0 ideas what is causing the issue...

LeagueOfPoro commented:
> Can someone elaborate on why the [BREAKING] was added to the title? As far as I'm aware, I cant find any conflicts with this and the current `master` branch...

It is breaking change - it requires new dependencies

Skribb11es committed:
Merge branch 'master' into master

Skribb11es committed:
Added check to circumvent late IMAP logins

Intended to prevent the event in which the IMAP session is created after the email is received from a riot by adding a timeout. Also added more exceptions + integrated it with all new commits to `master`.

Skribb11es committed:
New exceptions for IMAP

Skribb11es commented:
Fixed conflicts with the current `master` branch, along with adding some much-needed redundancy in the event that either the 2FA email arrived before the IMAP session could be opened, or the event that the 2FA email could not be found within the allotted time, to prevent hangups on waiting for an email that was either already received or never came.

Skribb11es committed:
Merge branch 'master' into master

auroraisluna commented on src/Browser.py:
```diff
@@ -132,11 +129,28 @@ def login(self, username: str, password: str, imapusername: str, imappassword: s
             # Currently unused but the call might be important server-side
             resPasToken = self.client.get(
                 "https://account.rewards.lolesports.com/v1/session/clientconfig/rms", headers=headers).close()
+            print("wow")
```
?

auroraisluna commented on src/Browser.py:
```diff
@@ -132,11 +129,28 @@ def login(self, username: str, password: str, imapusername: str, imappassword: s
             if resAccessToken.status_code == 200:
                 self.__dumpCookies()
                 return True
+            print("thats good")
```
?

Skribb11es replied:
This and above were both left in from testing, debug sadly does not give line by line reports otherwise these would not be needed.

Skribb11es replied:
Will be removed ASAP though

auroraisluna commented on src/Stats.py:
```diff
@@ -1,8 +1,7 @@
-    def __init__(self, farmThreads) -> None:
-        self.farmThreads = farmThreads
+    def __init__(self) -> None:
         self.accountData = {}
```
What's this change for?

LeagueOfPoro replied:
I agree with the removal. However, I don't think there really is a noticeable efficiency gain thanks to the way the assignment work. Some reading if you are interested: https://realpython.com/python-pass-by-reference/#passing-arguments-in-python

Skribb11es replied:
Overall it would be very minor, but if there is no need for it in the first place why have it...

Skribb11es committed:
Remove print statements

Skribb11es commented:
Should be the final build unless yall notice anything, as far as i can tell I've ironed out most of the bugs, and it should integrate just fine with the current master. The only thing that needs to be done from this point is probably IMAP docs for the wiki, of which I have never done before.

Skribb11es commented:
Finished writing docs/wiki page for IMAP integration.



A relatively simple integration of IMAP into [Capsule Farmer Evolved](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved), to automate 2FA.

This is by no means currently built to accept all IMAP connections and was designed to support most modern email providers. In the future, this may potentially allow for all IMAP connections, regardless of security, but for the current build, it only supports connections via SSL.

By `Skribb11es#0001`. Tested using Gmail.

# IMPORTANT!
Do note that by using automatic IMAP integration you are allowing the farmer to gain access to your email account for a brief period, this means that some emails, primarily if they are received during the time that the farmer is monitoring, will be "scanned". This means that the email will be marked as read, and the sender of the email processed. While this is as far as the farmer will go unless the email is from `noreply@mail.accounts.riotgames.com`, it does mean that you consent to give the farmer access to potentially private information.

Please understand that **NONE** of the information processed, during the period the farmer has access to your email, will be processed **anywhere** outside of **your computer**.

By `Skribb11es#0001`. Tested using Gmail.

# Setup
It is recommended that you use the [Config Helper](https://leagueofporo.com/confighelper.html), both for ease of use and because it makes configuring the farmer much easier to understand where and how to input the information.

Otherwise, we recommend using a dedicated text editor like [Notepad++](https://notepad-plus-plus.org/) or [Visual Studio Code](https://code.visualstudio.com/) in place of something like Notepad.

## How to find IMAP information
Almost every email provider has different policies when it comes to IMAP, for the sake of keeping this guide simple we are unable to list all the possible ways to set up IMAP for your email provider. However, if you need to locate this information, we highly recommend searching for "**[EMAIL PROVIDER HERE]** IMAP docs". This in almost all cases will provide you with official docs on how you can both enable and access IMAP in a different application, along with some more extensive configuration information.

As it stands, only Gmail is verified to work with the IMAP integration, if you encounter an issue with your particular email provider, please verify that the information you are inputting is correct and that you don't require any further security measures (in the case of Gmail you must use a dedicated App Password, and have 2FA enabled on your Google account). If you are still encountering an issue, please contact me `Skribb11es#0001` on Discord.

## Option 1 - Adding IMAP via Config Helper
You can use the config helper tool via the link provided [here](https://leagueofporo.com/confighelper.html), to create your config. Further information on how to locate the needed information can be found in the following steps below.

Assuming you have already made a config for your accounts and are looking to add automatic 2FA, it is recommended that you copy over all of the information for the accounts into the provided default fields (non-IMAP related fields), and then follow through with IMAP. For more information, please reference [App Configuration](Configuration).

1. Input your IMAP username and password into their respective fields `IMAP Username: (OPTIONAL)` and `IMAP Password: (OPTIONAL)`. If you are unsure what either of these may be, consult the documentation provided by your email service for IMAP logins. In most cases, it is your full email address + your account password (the password you use when you log in to your inbox). In the case of a more secure service like Gmail, it may be required for you to provide an App Password, instead of your default account password.

2. Locate the IMAP server for your email provider, this is usually located in the same docs as referenced above, and will be input into the `IMAP Server: (OPTIONAL)` field. Assuming the email provider you are using supports SSL connections this should be the end of your configuration.

3. Click **ADD ACCOUNT** once all the fields are filled out, and download or otherwise save the created config to the `config.yaml` file located in `./config/config.yaml`.

## Option 2 - Adding IMAP manually
If you are a more technical user, or just find it easier to add the necessary fields yourself you can do so by following the steps below.

1. Open `config.yaml` using your text editor of choice (ex. [Notepad++](https://notepad-plus-plus.org/))

2. Under each of the accounts you have set up add the following fields: 
```yaml
imapusername: ""
imappassword: ""
imapserver: ""
```
In total you should have something that resembles the following :
```yaml
accounts:
  accountname:
    username: ""
    password: ""
    imapusername: ""
    imappassword: ""
    imapserver: ""
```

3. Following the table below, fill out the new fields with their according information.

| Option | Description | Example | Default | Is required? |
| --- | --- | --- | --- | --- |
| `imapusername` | The specified username for logging in with IMAP according to the email provider's docs | thisisanexample@gmail.com | "" | Yes
| `imappassword` | The specified password for logging in with IMAP according to the email provider's docs | Password! | "" | Yes
| `imapserver` | The specified server to connect to according to the email provider's docs |imap.gmail.com | "" | Yes
Make sure to replace all variable values and to leave none blank, as this leaves potential for error, and the possibility that the farmer is not able to log into your IMAP account.

4. Save the config file and run the program.

# What to expect
Below is a list of possible outcomes when running the program after configuring IMAP. If you run into one of the listed errors, please consult the above guide again.
## Normal Operation
If you have properly configured all of the fields, the farmer will go ahead and log in to your Riot account via the specified username and password, at which point it will send a request to the specified IMAP server to log in. It will then wait until a new email is received, upon which it will check if the sender was `noreply@mail.accounts.riotgames.com`, and if it is then it will extract the 2FA code from the email's subject and store it. It will then close the IMAP session and log out, at which point it will then send the saved code to Riot's authentication servers, and log your account in. During this process, the status will only display `WAIT`. After the code has been fetched, the status will update to `LOGIN` and then `FETCHED 2FA CODE`, at which point it may stall for a little while it logs into the Riot servers. Once completed it will display `LIVE` next to the account name, and you should be fully up and running. 

## Failure to find 2FA code
In the case that a 2FA code is unable to be found by the farmer (this information is only displayed in the logs `./logs/capsulefarmer.log`), the farmer will retry after the allotted period, and attempt to find a 2FA code again. If you find that the farmer is consistently unable to find a 2FA code, please make sure that you are not blocking emails from Riot, and that you are using the correct email account for the specified Riot account. 


## Failure to log in 
If an error occurs logging into your email account via IMAP, the farmer will display `IMAP LOGIN FAILED` as the status, along with automatically stopping itself from ever retrying to log back into your Riot account, or email account, until you verify that your information is correct and restart the program. This is to insure that your account is not locked down due to multiple login requests without providing a 2FA code.

LeagueOfPoro commented on src/Config.py:
```diff
@@ -1,91 +1,193 @@
+            print("Press any key to exit...")
+            input()
+            raise ex
+
+        # Get bestStreams from URL
+        try:
+            remoteBestStreamsFile = requests.get(self.REMOTE_BEST_STREAMS_URL)
+            if remoteBestStreamsFile.status_code == 200:
+                self.bestStreams = remoteBestStreamsFile.text.split()
+        except Exception as ex:
+            print(f"[red]CRITICAL ERROR: Beststreams couldn't be loaded. Are you connected to the internet?")
+            print("Press any key to exit...")
+            input()
+            raise ex
+
+
+    def getAccount(self, account: str) -> dict:
+        """
+        Get account information
+
+        :param account: string, name of the account
+        :return: dictionary, account information
+        """
+        return self.accounts[account]
+
+    def __findConfig(self, configPath):
+        """
+        Try to find configuartion file in alternative locations.
+
+        :param configPath: user suplied configuartion file path
+        :return: pathlib.Path, path to the configuration file
+        """
+        configPath = Path(configPath)
+        if configPath.exists():
+            return configPath
+        if Path("../config/config.yaml").exists():
+            return Path("../config/config.yaml")
+        if Path("config/config.yaml").exists():
+            return Path("config/config.yaml")
+        return configPath
```
Duplicate class

LeagueOfPoro replied:
This is the correct class, delete the preceding one.

LeagueOfPoro commented on src/Browser.py:
```diff
@@ -64,15 +71,30 @@ def login(self, username: str, password: str, refreshLock) -> bool:
-                    "https://auth.riotgames.com/api/v1/authorization", json=data)
-                resJson = res.json()
+                if (imapserver != ""):
+                    #Handles all IMAP requests
```
Release the lock to enable UI updates
```python
refreshLock.release()
self.stats.updateStatus(self.account, f"[green]WAITING 2FA EMAIL")
```

And then in finally:
```python
if refreshLock.locked():
   refreshLock.release()
```

Skribb11es committed:
Fixed issues, as per request of Poro

Skribb11es commented:
Done and dusted, not sure how the config managed to be duplicated in the same file, but it has since been fixed.

LeagueOfPoro commented on src/Browser.py:
```diff
@@ -129,9 +131,11 @@ def login(self, username: str, password: str, imapusername: str, imappassword: s
             if resAccessToken.status_code == 200:
                 self.__dumpCookies()
                 return True
+            print("thats good")
```
...

LeagueOfPoro commented on src/Browser.py:
```diff
@@ -129,9 +131,11 @@ def login(self, username: str, password: str, imapusername: str, imappassword: s
             # Currently unused but the call might be important server-side
             resPasToken = self.client.get(
                 "https://account.rewards.lolesports.com/v1/session/clientconfig/rms", headers=headers).close()
+            print("wow")
```
...

Skribb11es commented:
Forgot to remove those lines from my local build 🤦‍♂️

Skribb11es committed:
Fixed API "crashing" + removed print statements

see PR [#154](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/pull/154)

Skribb11es commented:
PR #154 for the changes made by the API fix

Skribb11es committed:
Fixed case issue

Skribb11es committed:
Refresh fix + IMAP timeout fix

Skribb11es committed:
Delete InvalidIMAPCredentials.py

Skribb11es committed:
Add files via upload

Skribb11es committed:
Merge branch 'master' into master

LeagueOfPoro commented:
Before I merge it, please look at your git settings and make sure you have `autocrlf` enabled. If you had it disabled, commit all files except the `main.py` again. I would prefer not to have the files show up as completely changed in case there is a problem.

More info: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

Skribb11es committed:
Merge pull request #2 from LeagueOfPoro/master

d

Skribb11es committed:
Line Endings

Skribb11es committed:
Fixed line endings

Skribb11es committed:
Remove print statement

LeagueOfPoro commented on src/Browser.py:
```diff
@@ -55,7 +63,6 @@ def login(self, username: str, password: str, refreshLock) -> bool:
             "https://login.leagueoflegends.com/?redirect_uri=https://lolesports.com/&lang=en")
         self.__loadCookies()
         try:
-            refreshLock.acquire()
```
This was supposed to stay here. Now the lock is never locked, and the farmer crashes every time it tries to use imap. I will fix it.

LeagueOfPoro committed:
Lock the UI updates

TheDingo8MyBaby commented:
Does anyone know how to get this work with ProtonMail ProtonBridge?

Imap Server isn't like gmail, it's a hostname/port

Hostname: 127.0.0.1
Port: 1143

tried `imapServer: "127.0.0.1:1143"` and got 
```IMAP LOGIN FAILED```
